plugin-server/src/worker/ingestion/event-pipeline/embedErrorEventStep.ts

```diff
@@ -0,0 +1,92 @@
+import { FeatureExtractionPipeline, FeatureExtractionPipelineOptions } from '@xenova/transformers'
+
+import { defaultConfig } from '../../../config/config'
+import { runInstrumentedFunction } from '../../../main/utils'
+import { PreIngestionEvent } from '../../../types'
+import { status } from '../../../utils/status'
+import { EventPipelineRunner } from './runner'
+
+const errorEmbeddingModel = defaultConfig.ERROR_EVENT_EMBEDDING_MODEL
+const errorEmbeddingEnabledTeams = (defaultConfig.ENABLE_ERROR_EMBEDDING_TEAM_IDS || '')
+    .split(',')
+    .map((id) => parseInt(id, 10))
+
+let featureExtractionPipeline: FeatureExtractionPipeline | null
+
+// this downloads the model the first time it is run, so eventually we'll do this on server start-up
+// for now we only want to spend the cost of loading the model if we are actually using it
+export async function initEmbeddingModel(): Promise<FeatureExtractionPipeline | null> {
+    try {
+        // if no teams are embedding events, we don't need to load the model
+        // this lets us release this and opt in to timing it in production
+        const anyTeamsAreEmbeddingEvents = errorEmbeddingEnabledTeams.length > 0
+
+        if (!featureExtractionPipeline && anyTeamsAreEmbeddingEvents) {
+            // a little magic here to both delay the slow import until the first time it is needed
+            // and to make it work due to some ESM/commonjs faff
+            await runInstrumentedFunction({
+                func: async () => {
+                    const TransformersApi = Function('return import("@xenova/transformers")')()
+                    const { pipeline } = await TransformersApi
+                    featureExtractionPipeline = await pipeline('feature-extraction', errorEmbeddingModel)
+                },
+                statsKey: 'initErrorEventEmbeddingModel',
+            })
+        }
+    } catch (e) {
+        status.error('??', 'Error initializing error event embedding model', e)
+    }
+    return Promise.resolve(featureExtractionPipeline)
+}
+
+export async function embedErrorEvent(
```

Comment:
Is this function completely re-entrant? Is `feature-extraction` acting with a state that will be mutated, or completely stateless?

`plugins-ingestion` will run this tasks several times in parallel, up to `INGESTION_CONCURRENCY`.

Reply:
yep, or at least the docs make it seem that way 👍